### PR TITLE
Remove wp-data & wc-blocks-checkout dependencies from OA script

### LINKED ISSUE #43829: Fix unnecessary JS script loading on the frontend related to Order Attribution Feature.
### Prerequisites

- [X] I have carried out troubleshooting steps and I believe I have found a bug.
- [X] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

This [support forum thread](https://wordpress.org/support/topic/more-than-39-js-files-are-loaded-on-version-8-5-1/) reports the loading of an additional 39 scripts beginning with WooCommerce 8.5/.1. On initial review, it looks like it may have been introduced [via a dependency on the `wc-blocks-checkout` and `wp-data` scripts](https://github.com/woocommerce/woocommerce/blob/7f2fd37fad64c890a51293a49661d1d166309c32/plugins/woocommerce/src/Internal/Orders/OrderAttributionController.php#L231) in the new `wc-order-attribution` feature. If that is enqueued in every page load that is definitely, the source for the extra scripts and should be rectified.

### Expected behavior

WooCommerce should not be introducing a significant payload increase unnecessarily on every page load.

### Actual behavior

Significant payload increase is present.

### Steps to reproduce

Install and activate WooCommerce 8.5 or 8.5.1 with the Order Attribution feature active (it is active by default).

### WordPress Environment

Doesn't matter.

### Isolating the problem

- [ ] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [ ] This bug happens with a default WordPress theme active, or [Storefront](https://woo.com/products/storefront/).
- [ ] I can reproduce this bug consistently using the steps above.

### PR DESCRIPTION
**Pull request by tomalec**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

- Remove wp-data & wc-blocks-checkout dependencies from OA script
   Resolve conditional dependency in JS.

Closes: https://github.com/woocommerce/woocommerce/issues/43829

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

0. (developer only) Check that order-attribution js does not have `'wp-data', 'wc-blocks-checkout'` as dependencies.
1. Check live site:
   1. Ensure the mini-cart block is removed from your header.
      image

   2. Load the page and ensure the `wc-blocks-checkout` script is not loaded.
      Call from the browser dev tools `document.querySelector('script#wc-blocks-checkout-js')` - should return `null`
2. Check testing instructions for the following for the regression:
   1. https://github.com/woocommerce/woocommerce/pull/39701
   3. https://github.com/woocommerce/woocommerce/pull/43012

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce/client/legacy/js/frontend/order-attribution.js [hunk 1]
```diff
@@ -111,4 +114,28 @@
...
+	// Conditionally, wait for and use 'wp-data' & 'wc-blocks-checkout.
+
+	// Wait for (async) block checkout initialization and set source values once loaded.
+	function eventuallyInitializeCheckoutBlock() {
+		if (
+			window.wp && window.wp.data && typeof window.wp.data.subscribe === 'function'
+		) {
+			const unsubscribe = wp.data.subscribe( function () {
+				const checkoutDataStore = wp.data.select( CHECKOUT_STORE_KEY );
+				if ( undefined !== checkoutDataStore && checkoutDataStore.getOrderId() !== 0 ) {
```
Comment (opr) [inline plugins/woocommerce/client/legacy/js/frontend/order-attribution.js]:
```suggestion
				if ( 'undefined' !== typeof checkoutDataStore && checkoutDataStore.getOrderId() !== 0 ) {
```

I prefer this way of doing it as it's a little safer.

**Review by opr — CHANGES_REQUESTED** [review]
From the blocks point of view this is working nicely.

Just to recap: I removed the mini cart block from my site and tested before this PR. There were 55 JS scripts loaded, `wc-blocks-checkout` was one of them. The only dependent listed for this was `wc-order-attribution`. `wp-data` was also loaded.

After applying this PR, the total number of scripts loaded was 13, `wc-blocks-checkout` and `wp-data` were not included.

Do you think updating the testing instructions to include something like:

1. Ensure the mini-cart block is removed from your header.
2. Load the page and ensure the `wc-blocks-checkout` script is not loaded.

would be useful? Maybe mark it as "developer only" instructions?

File: plugins/woocommerce/client/legacy/js/frontend/order-attribution.js (hunk 1 above)
Comment (tomalec) [reply plugins/woocommerce/client/legacy/js/frontend/order-attribution.js]: `checkoutDataStore` constant is declared just the line above, so IMHO typeof is not needed.

Also, do we even need that check? We subscribed to `CHECKOUT_STORE_KEY` store changes, so `CHECKOUT_STORE_KEY` store should be guaranteed to exist in the callback, shouldn't it?

Comment (tomalec) [reply plugins/woocommerce/client/legacy/js/frontend/order-attribution.js]: I did a quick micro-perf check, and it seems there is no significant difference between them. https://jsperf.app/nofavo

Using `typeof` here would make lines 127 and below "safe" in the sense they will run without visible error when line 126 is lost (for example, in a bad merge).
IMHO, this is actually **not desired**. If the line 126 is lost, the code should fail fast. So we could identify the bad merge or other bug before shipping it to production.

**Comment by tomalec** [comment]
Thank you for the fast review and testing.

I added the testing instructions as you suggested.

File: plugins/woocommerce/client/legacy/js/frontend/order-attribution.js (hunk 1 above)
Comment (tomalec) [reply plugins/woocommerce/client/legacy/js/frontend/order-attribution.js]: I removed this condition altogether in https://github.com/woocommerce/woocommerce/pull/43857/commits/f3abaf0304f75d5a0edf2a4e7709cd344ef4bfd3

We don't need to check if there is an order being processed in the checkout.
To add extension data, we only need the store to finally be there. That should be guaranteed by the  `subscribe` itself.

Comment (opr) [reply plugins/woocommerce/client/legacy/js/frontend/order-attribution.js]: @tomalec I appreciate how deeply you considered this! Kudos. 🙏🏼 The result is better for sure.

**Review by opr — APPROVED** [review]
Thanks for adding the test instructions, could we just specify checking the script dependencies is "dev only" so globalstep don't have to figure out how to do that?

**Comment by tomalec** [comment]
> could we just specify checking the script dependencies is "dev only" 

Do you mean updating the _"Check that order-attribution js does not have 'wp-data', 'wc-blocks-checkout' as dependencies."_? if so - done.

**Comment by tomalec** [comment]
Speaking of being super safe I've added [one more tweak](https://github.com/woocommerce/woocommerce/pull/43857/commits/5685035e694bc22378a4fccdefbf9d63b4c600b4)  to use `window.wp.` instead of `wp.` 
It was still after checking that `window.wp` is there, but it's always safer to use the same thing and be kind to linters and bundlers, as the file would use only one global.

**Comment by tjcafferkey** [comment]
I've done some testing:

* The steps in this PR test well 👍🏻 
* https://github.com/woocommerce/woocommerce/pull/39701 tests well with no regressions for me 👍🏻 
* https://github.com/woocommerce/woocommerce/pull/43012 doesn't test well for me, perhaps I'm missing something but I've put some notes below 👎🏻 

**Fix Consent API integration with OSA (trunk-based)**

"Denied by default" step didn't return `false` values for:

>wc_order_attribution.params.allowTracking should be set to false and no order attribution data should be collected.

Screenshot 2024-01-24 at 07 19 33

I also didn't have a `/checkout` request for the checkout block steps, for some reason. Perhaps I'm doing something obviously wrong but I've also asked about this in the Team Rubik channel. I stopped testing this PR until I've resolved these issues if its user-error on my end.

**Comment by tomalec** [comment]
Thank you, @tjcafferkey, for the testing.

> "Denied by default" step didn't return false values for:

Could you share your Complianz settings? I'd rather avoid merging if there's a chance to release another bug in regards of tracking.

> I also didn't have a `/checkout` request for the checkout block steps,

The checkout request should happen after you place an order.


Do you have both  WP Consent API and Complianz plugins installed?

**Comment by tomalec** [comment]
I replicated the `opt-in` + `wc_order_attribution.params.allowTracking == true` with Complianz set to UK mode, and with WP Consent API plugin deactivated.
image

If that's the case for you. Please install WP Consent API plugin, as it's required for the integration.

**Comment by tomalec** [comment]
If the problem is with `wc_order_attribution.params.allowTracking == true`, then it's most likely something unrelated to the changes from this PR, as none of the changed lines affect the flow that sets `allowTracking`.

I'm merging this PR to fix https://github.com/woocommerce/woocommerce/issues/43829 in the upcoming release. 
If the problem with `allowTracking` still happens with the WP Consent API installed, then it could be tracked as a separate issue.

**Pull request merged**